WalletWasabi/Crypto/Api/WabiSabiClient.cs

```diff
@@ -0,0 +1,220 @@
+
+		public List<Credential> Credentials { get; } = new List<Credential>();
+
+		public void HandleResponse(RegistrationResponse registrationResponse)
+		{
+			Guard.NotNull(nameof(registrationResponse), registrationResponse);
+			
+			if (Transcript is null)
+			{
+				throw new InvalidOperationException("No transcript was initialised.");
+			}
+
+			var requestedCredentials = CredentialsRequested;
+			var issuedCredentials = registrationResponse.IssuedCredentials;
+
+			var issuedCredentialCount = issuedCredentials.Count();
+			var requestedCredentialCount = requestedCredentials.Count();
+			if (issuedCredentialCount != NumberOfCredentials)
+			{
+				throw new WabiSabiException(
+					WabiSabiErrorCode.IssuedCredentialNumberMismatch, 
+					$"{issuedCredentialCount} issued but {requestedCredentialCount} were requested.");
+			}
+
+			var credentials = Enumerable
+				.Zip(requestedCredentials, issuedCredentials)
+				.Select(x => new { Requested = x.First, Issued = x.Second })
+				.ToArray();
+
+			var statements = credentials
+				.Select(x => ProofSystem.IssuerParameters(CoordinatorParameters, x.Issued, x.Requested.Credential.Ma));
+
+			var areCorrectlyIssued = Verifier.Verify(Transcript, statements, registrationResponse.Proofs);
+			if (!areCorrectlyIssued)
+			{
+				throw new WabiSabiException(WabiSabiErrorCode.ClientReceivedInvalidProofs);
+			}
+
+			foreach (var presented in CredentialsToPresent)
+			{
+				Credentials.Remove(presented.Credential);
+			}
+
+			CredentialsToPresent.Clear();
+			CredentialsRequested.Clear();
+			CredentialsToRequest.Clear();
+
+			var credentialReceived = credentials.Select(x => 
+				new Credential(new Scalar((ulong)x.Requested.Amount.Satoshi), x.Requested.r, x.Issued));
+			Credentials.AddRange(credentialReceived);
+		}
+
+		public ICredentialRequestBuilder AsCredentialRequestBuilder()
+		{
+			return this;
+		} 
+
+		ICredentialRequestBuilder ICredentialRequestBuilder.RequestCredentialFor(Money amount)
+		{
+			Guard.NotNull(nameof(amount), amount);
+			Guard.InRangeAndNotNull(nameof(amount), amount, Money.Zero, Constants.MaximumCredentailAmount);
+
+			CredentialsToRequest.Add(amount);
+
+			return this;
+		}
+
+		ICredentialRequestBuilder ICredentialRequestBuilder.PresentCredentials(params Credential[] credentials)
+		{
+			Guard.NotNullOrEmpty(nameof(credentials), credentials);
+
+			foreach (var credential in credentials)
+			{
+				var z = Random.GetScalar();
+				if (!CredentialsToPresent.Any(x => x.Credential.Mac == credential.Mac))
+				{
+					var knowledge = ProofSystem.ShowCredential(credential.Present(z), z, credential, CoordinatorParameters);
+					CredentialsToPresent.Add((credential, z, knowledge));
+				}
+			}
+			return this;
+		}
+
+		RegistrationRequest ICredentialRequestBuilder.Build()
+		{
+			// Make sure we request always the same number of credentials
+			var missingCredentialRequests = NumberOfCredentials - CredentialsToRequest.Count();
```

Comment:
Super minor Nits: I would rename `missingCredentialRequests` to maybe `zeroValueCredentialRequests` or `nullValueCredentialRequests`. Those credentials are not "missing"....they just have zero/null values.